python/ray/dag/dag_node.py

```diff
@@ -380,9 +386,20 @@ def traverse_and_apply(self, fn: "Callable[[DAGNode], T]"):
         """
         visited = set()
         queue = [self]
+        num_output_nodes = 0
+
         while queue:
             node = queue.pop(0)
             if node not in visited:
+                if node.is_output_node:
+                    num_output_nodes += 1
+                    # Validate whether there are multiple nodes that call
+                    # `experimental_compile`.
+                    if num_output_nodes > 1:
+                        raise ValueError(
```

Comment:
do we need this? we already have a check in the highest layer. can we just assert instead?

Reply:
What's the highest layer that you are referring to?

In `experimental_compile`, we checked whether the "same" DAG node calls `experimental_compile` multiple times. In `traverse_and_apply`, it checks whether there are "multiple" DAG nodes that call `experimental_compile`.

Reply:
is it possible we face this error if we already raise an error when the same dag calls `experimental_compile ` twice? If no, I think it should be assert, if yes, can you give me an example?

Reply:
By following @ruisearch42's suggestions, I eventually used `adag_output_node` instead of `num_output_nodes`.